Shared EntityManager exception when updating with an EnittyView

### Description
Creating an an `Entity` (using a JPA repository ) and trying to update it using an `EntityView` with the `EntityViewManager`

**Code/POC**

- https://github.com/EugenMayer/java-spring-pocs/tree/master/src/main/java/de/kontextwork/poc/spring/blaze
- https://github.com/EugenMayer/java-spring-pocs/blob/master/src/test/java/de/kontextwork/poc/spring/blaze/BlazeServiceTest.java

run it using

```
./gradlew bpTest --tests BlazeServiceTest.updateUserProfile
```

You will need to have the current master/HEAD build of BP, change the snapshot name accordingly to your local one at 

https://github.com/EugenMayer/java-spring-pocs/blob/master/build.gradle#L41

### Actual behavior
An exception is throwm

```
  java.lang.IllegalStateException: Could not access entity transaction!
      at de.kontextwork.poc.spring.blaze.BlazeServiceTest.updateUserProfile(BlazeServiceTest.java:83)
  Caused by: java.lang.IllegalStateException: Not allowed to create transaction on shared EntityManager - use Spring transactions or EJB CMT instead
      at de.kontextwork.poc.spring.blaze.BlazeServiceTest.updateUserProfile(BlazeServiceTest.java:83)
```


### Environment
Version: master HEAD on `f6405863e`   				
PA-Provider:       Hibernate 5.4.5
DBMS:               h2 / mariadb
Application Server: spring boot 2.1.9

--- a/.travis-deploy.sh
+++ b/.travis-deploy.sh
@@ -8,7 +8,7 @@ if [ "$TRAVIS_REPO_SLUG" == "Blazebit/blaze-persistence" ] &&
     [ "x$JDK" == "x" ]; then
 
   echo "Starting snapshot deployment..."
-  mvn -s .travis-settings.xml -DperformRelease -DskipTests -Dgpg.skip=true -Dquiet=true deploy
+  mvn -s .travis-settings.xml -DperformRelease -DskipTests -Dgpg.skip=true -Dquiet=true clean deploy
   echo "Snapshots deployed!"
 
 else

--- a/entity-view/api/src/main/java/com/blazebit/persistence/view/spi/TransactionAccess.java
+++ b/entity-view/api/src/main/java/com/blazebit/persistence/view/spi/TransactionAccess.java
@@ -1,0 +1,48 @@
+/*
+ * Copyright 2014 - 2019 Blazebit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.blazebit.persistence.view.spi;
+
+import javax.transaction.Synchronization;
+
+/**
+ * A transaction control abstraction that allows to be independent of the concrete transaction technology.
+ *
+ * @author Christian Beikov
+ * @since 1.4.0
+ */
+public interface TransactionAccess {
+
+    /**
+     * Returns whether the current transaction is active.
+     *
+     * @return true if active, false otherwise
+     */
+    public boolean isActive();
+
+    /**
+     * Mark the current transaction as rollback only.
+     */
+    public void markRollbackOnly();
+
+    /**
+     * Registers the given synchronization in the current transaction.
+     *
+     * @param synchronization The synchronization to register
+     */
+    public void registerSynchronization(Synchronization synchronization);
+    
+}

--- a/entity-view/api/src/main/java/com/blazebit/persistence/view/spi/TransactionAccessFactory.java
+++ b/entity-view/api/src/main/java/com/blazebit/persistence/view/spi/TransactionAccessFactory.java
@@ -14,21 +14,25 @@
  * limitations under the License.
  */
 
-package com.blazebit.persistence.view.impl.tx;
+package com.blazebit.persistence.view.spi;
 
-import javax.transaction.Synchronization;
+import javax.persistence.EntityManager;
 
 /**
+ * A factory for creating a {@link TransactionAccess}.
+ * This is created via the {@link java.util.ServiceLoader} API.
  *
  * @author Christian Beikov
- * @since 1.2.0
+ * @since 1.4.0
  */
-public interface TransactionSynchronizationStrategy {
+public interface TransactionAccessFactory {
 
-    public boolean isActive();
-
-    public void markRollbackOnly();
-    
-    public void registerSynchronization(Synchronization synchronization);
+    /**
+     * Creates a transaction access object.
+     *
+     * @param entityManager The entity manager associated with the transaction.
+     * @return The transaction access object
+     */
+    public TransactionAccess createTransactionAccess(EntityManager entityManager);
     
 }

--- a/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/EntityViewManagerImpl.java
+++ b/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/EntityViewManagerImpl.java
@@ -414,7 +414,7 @@ public class EntityViewManagerImpl implements EntityViewManager {
                 updater.remove(context, proxy);
             }
         } catch (Throwable t) {
-            context.getSynchronizationStrategy().markRollbackOnly();
+            context.getTransactionAccess().markRollbackOnly();
             ExceptionUtils.doThrow(t);
         }
     }
@@ -430,7 +430,7 @@ public class EntityViewManagerImpl implements EntityViewManager {
         try {
             updater.remove(context, viewId);
         } catch (Throwable t) {
-            context.getSynchronizationStrategy().markRollbackOnly();
+            context.getTransactionAccess().markRollbackOnly();
             ExceptionUtils.doThrow(t);
         }
     }
@@ -455,7 +455,7 @@ public class EntityViewManagerImpl implements EntityViewManager {
                 updater.executeUpdate(context, updatableProxy);
             }
         } catch (Throwable t) {
-            context.getSynchronizationStrategy().markRollbackOnly();
+            context.getTransactionAccess().markRollbackOnly();
             ExceptionUtils.doThrow(t);
         }
     }

--- a/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/tx/Hibernate4TransactionSynchronizationStrategy.java
+++ b/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/tx/Hibernate4TransactionSynchronizationStrategy.java
@@ -16,21 +16,21 @@
 
 package com.blazebit.persistence.view.impl.tx;
 
-import java.lang.reflect.Method;
+import com.blazebit.persistence.view.spi.TransactionAccess;
+import com.blazebit.reflection.ExpressionUtils;
+import com.blazebit.reflection.ReflectionUtils;
 
 import javax.persistence.EntityManager;
 import javax.persistence.EntityTransaction;
 import javax.transaction.Synchronization;
-
-import com.blazebit.reflection.ExpressionUtils;
-import com.blazebit.reflection.ReflectionUtils;
+import java.lang.reflect.Method;
 
 /**
  *
  * @author Christian Beikov
  * @since 1.2.0
  */
-public class Hibernate4TransactionSynchronizationStrategy implements TransactionSynchronizationStrategy {
+public class Hibernate4TransactionSynchronizationStrategy implements TransactionAccess {
     
     private final EntityTransaction tx;
     private final Object synchronizationRegistry;
@@ -64,12 +64,7 @@ public class Hibernate4TransactionSynchronizationStrategy implements Transaction
     @Override
     public void registerSynchronization(Synchronization synchronization) {
         try {
-            SynchronizationRegistry registry = SynchronizationRegistry.getRegistry(synchronizationRegistry);
-            if (registry == null) {
-                registry = new SynchronizationRegistry(synchronizationRegistry);
-                registerSynchronization.invoke(synchronizationRegistry, registry);
-            }
-            registry.addSynchronization(synchronization);
+            registerSynchronization.invoke(synchronizationRegistry, synchronization);
         } catch (Exception e) {
             throw new RuntimeException(e);
         }

--- a/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/tx/Hibernate5EntityTransactionSynchronizationStrategy.java
+++ b/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/tx/Hibernate5EntityTransactionSynchronizationStrategy.java
@@ -16,21 +16,21 @@
 
 package com.blazebit.persistence.view.impl.tx;
 
-import java.lang.reflect.Method;
+import com.blazebit.persistence.view.spi.TransactionAccess;
+import com.blazebit.reflection.ExpressionUtils;
+import com.blazebit.reflection.ReflectionUtils;
 
 import javax.persistence.EntityManager;
 import javax.persistence.EntityTransaction;
 import javax.transaction.Synchronization;
-
-import com.blazebit.reflection.ExpressionUtils;
-import com.blazebit.reflection.ReflectionUtils;
+import java.lang.reflect.Method;
 
 /**
  *
  * @author Christian Beikov
  * @since 1.2.0
  */
-public class Hibernate5EntityTransactionSynchronizationStrategy implements TransactionSynchronizationStrategy {
+public class Hibernate5EntityTransactionSynchronizationStrategy implements TransactionAccess {
     
     private final EntityTransaction tx;
     private final Object synchronizationRegistry;
@@ -64,12 +64,7 @@ public class Hibernate5EntityTransactionSynchronizationStrategy implements Trans
     @Override
     public void registerSynchronization(Synchronization synchronization) {
         try {
-            SynchronizationRegistry registry = SynchronizationRegistry.getRegistry(synchronizationRegistry);
-            if (registry == null) {
-                registry = new SynchronizationRegistry(synchronizationRegistry);
-                registerSynchronization.invoke(synchronizationRegistry, registry);
-            }
-            registry.addSynchronization(synchronization);
+            registerSynchronization.invoke(synchronizationRegistry, synchronization);
         } catch (Exception e) {
             throw new RuntimeException(e);
         }

--- a/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/tx/Hibernate5JtaPlatformTransactionSynchronizationStrategy.java
+++ b/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/tx/Hibernate5JtaPlatformTransactionSynchronizationStrategy.java
@@ -16,6 +16,7 @@
 
 package com.blazebit.persistence.view.impl.tx;
 
+import com.blazebit.persistence.view.spi.TransactionAccess;
 import com.blazebit.reflection.ReflectionUtils;
 
 import javax.transaction.Status;
@@ -27,7 +28,7 @@ import java.lang.reflect.Method;
  * @author Moritz Becker
  * @since 1.4.0
  */
-public class Hibernate5JtaPlatformTransactionSynchronizationStrategy implements TransactionSynchronizationStrategy {
+public class Hibernate5JtaPlatformTransactionSynchronizationStrategy implements TransactionAccess {
 
     private final Object jtaPlatform;
     private final Object jtaTransactionManager;

--- a/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/tx/JtaTransactionSynchronizationStrategy.java
+++ b/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/tx/JtaTransactionSynchronizationStrategy.java
@@ -16,6 +16,8 @@
 
 package com.blazebit.persistence.view.impl.tx;
 
+import com.blazebit.persistence.view.spi.TransactionAccess;
+
 import javax.transaction.Status;
 import javax.transaction.Synchronization;
 import javax.transaction.TransactionSynchronizationRegistry;
@@ -25,7 +27,7 @@ import javax.transaction.TransactionSynchronizationRegistry;
  * @author Christian Beikov
  * @since 1.2.0
  */
-public class JtaTransactionSynchronizationStrategy implements TransactionSynchronizationStrategy {
+public class JtaTransactionSynchronizationStrategy implements TransactionAccess {
     
     private final TransactionSynchronizationRegistry synchronizationRegistry;
 
@@ -45,12 +47,7 @@ public class JtaTransactionSynchronizationStrategy implements TransactionSynchro
 
     @Override
     public void registerSynchronization(Synchronization synchronization) {
-        SynchronizationRegistry registry = (SynchronizationRegistry) synchronizationRegistry.getResource(SynchronizationRegistry.class.getName());
-        if (registry == null) {
-            registry = new SynchronizationRegistry();
-            synchronizationRegistry.registerInterposedSynchronization(registry);
-        }
-        registry.addSynchronization(synchronization);
+        synchronizationRegistry.registerInterposedSynchronization(synchronization);
     }
 
 }

--- a/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/tx/SynchronizationRegistry.java
+++ b/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/tx/SynchronizationRegistry.java
@@ -16,6 +16,8 @@
 
 package com.blazebit.persistence.view.impl.tx;
 
+import com.blazebit.persistence.view.spi.TransactionAccess;
+
 import javax.transaction.Status;
 import javax.transaction.Synchronization;
 import java.util.ArrayList;
@@ -29,7 +31,7 @@ import java.util.concurrent.ConcurrentMap;
  * @author Christian Beikov
  * @since 1.3.0
  */
-public class SynchronizationRegistry implements Synchronization {
+public class SynchronizationRegistry implements Synchronization, TransactionAccess {
 
     private static final ThreadLocal<Object> THREAD_LOCAL_KEY = new ThreadLocal<Object>() {
         @Override
@@ -38,24 +40,38 @@ public class SynchronizationRegistry implements Synchronization {
         }
     };
     private static final ConcurrentMap<Object, SynchronizationRegistry> REGISTRY = new ConcurrentHashMap<>();
+    private final TransactionAccess transactionAccess;
     private final List<Synchronization> synchronizations;
     private final Object key;
 
-    public SynchronizationRegistry() {
-        this(THREAD_LOCAL_KEY.get());
+    public SynchronizationRegistry(TransactionAccess transactionAccess) {
+        this(transactionAccess, THREAD_LOCAL_KEY.get());
     }
 
-    public SynchronizationRegistry(Object key) {
+    public SynchronizationRegistry(TransactionAccess transactionAccess, Object key) {
+        this.transactionAccess = transactionAccess;
         this.synchronizations = new ArrayList<>(1);
         this.key = key;
+        transactionAccess.registerSynchronization(this);
         REGISTRY.put(key, this);
     }
 
-    public static SynchronizationRegistry getRegistry(Object key) {
-        return REGISTRY.get(key);
+    public static SynchronizationRegistry getRegistry() {
+        return REGISTRY.get(THREAD_LOCAL_KEY.get());
     }
 
-    public void addSynchronization(Synchronization synchronization) {
+    @Override
+    public boolean isActive() {
+        return transactionAccess.isActive();
+    }
+
+    @Override
+    public void markRollbackOnly() {
+        transactionAccess.markRollbackOnly();
+    }
+
+    @Override
+    public void registerSynchronization(Synchronization synchronization) {
         synchronizations.add(synchronization);
     }
 

--- a/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/tx/TransactionHelper.java
+++ b/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/tx/TransactionHelper.java
@@ -16,6 +16,8 @@
 
 package com.blazebit.persistence.view.impl.tx;
 
+import com.blazebit.persistence.view.spi.TransactionAccess;
+import com.blazebit.persistence.view.spi.TransactionAccessFactory;
 import com.blazebit.reflection.ReflectionUtils;
 
 import javax.naming.InitialContext;
@@ -25,6 +27,9 @@ import javax.naming.NoInitialContextException;
 import javax.persistence.EntityManager;
 import javax.transaction.TransactionSynchronizationRegistry;
 import java.lang.reflect.InvocationTargetException;
+import java.util.ServiceLoader;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
 /**
  *
@@ -33,6 +38,7 @@ import java.lang.reflect.InvocationTargetException;
  */
 public class TransactionHelper {
 
+    private static final Logger LOG = Logger.getLogger(TransactionHelper.class.getName());
     private static final String[] NAMES = {
         "java:comp/TransactionSynchronizationRegistry", // Java EE Standard name
         "java:/TransactionSynchronizationRegistry", // Local providers
@@ -42,7 +48,16 @@ public class TransactionHelper {
     private TransactionHelper() {
     }
 
-    public static TransactionSynchronizationStrategy getSynchronizationStrategy(EntityManager em) {
+    public static TransactionAccess getTransactionAccess(EntityManager em) {
+        SynchronizationRegistry registry = SynchronizationRegistry.getRegistry();
+        if (registry == null) {
+            TransactionAccess transactionAccess = getTransactionAccessInternal(em);
+            registry = new SynchronizationRegistry(transactionAccess);
+        }
+        return registry;
+    }
+
+    private static TransactionAccess getTransactionAccessInternal(EntityManager em) {
         InitialContext context = null;
 
         try {
@@ -63,6 +78,17 @@ public class TransactionHelper {
                 } catch (NamingException e) {
                     throw new IllegalArgumentException("Could not access transaction synchronization registry!", e);
                 }
+            }
+        }
+
+        for (TransactionAccessFactory transactionAccessFactory : ServiceLoader.load(TransactionAccessFactory.class)) {
+            try {
+                TransactionAccess transactionAccess = transactionAccessFactory.createTransactionAccess(em);
+                if (transactionAccess != null) {
+                    return transactionAccess;
+                }
+            } catch (Exception ex) {
+                LOG.log(Level.WARNING, "Error during creation of transaction access!", ex);
             }
         }
 

--- a/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/DefaultUpdateContext.java
+++ b/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/DefaultUpdateContext.java
@@ -18,7 +18,7 @@ package com.blazebit.persistence.view.impl.update;
 
 import com.blazebit.persistence.view.impl.EntityViewManagerImpl;
 import com.blazebit.persistence.view.impl.tx.TransactionHelper;
-import com.blazebit.persistence.view.impl.tx.TransactionSynchronizationStrategy;
+import com.blazebit.persistence.view.spi.TransactionAccess;
 import com.blazebit.persistence.view.impl.update.flush.PostFlushDeleter;
 
 import javax.persistence.EntityManager;
@@ -40,7 +40,7 @@ public class DefaultUpdateContext implements UpdateContext {
     private final EntityViewManagerImpl evm;
     private final EntityManager em;
     private final boolean forceFull;
-    private final TransactionSynchronizationStrategy synchronizationStrategy;
+    private final TransactionAccess transactionAccess;
     private final InitialStateResetter initialStateResetter;
     private Map<Object, Object> removedObjects;
     private Set<EntityKey> versionChecked;
@@ -50,14 +50,14 @@ public class DefaultUpdateContext implements UpdateContext {
         this.evm = evm;
         this.em = em;
         this.forceFull = forceFull;
-        this.synchronizationStrategy = TransactionHelper.getSynchronizationStrategy(em);
+        this.transactionAccess = TransactionHelper.getTransactionAccess(em);
 
-        if (!synchronizationStrategy.isActive()) {
+        if (!transactionAccess.isActive()) {
             throw new IllegalStateException("Transaction is not active!");
         }
 
         this.initialStateResetter = new ResetInitialStateSynchronization();
-        synchronizationStrategy.registerSynchronization((Synchronization) initialStateResetter);
+        transactionAccess.registerSynchronization((Synchronization) initialStateResetter);
     }
 
     @Override
@@ -101,8 +101,8 @@ public class DefaultUpdateContext implements UpdateContext {
         return removedObjects != null && removedObjects.containsKey(value);
     }
 
-    public TransactionSynchronizationStrategy getSynchronizationStrategy() {
-        return synchronizationStrategy;
+    public TransactionAccess getTransactionAccess() {
+        return transactionAccess;
     }
 
     @Override

--- a/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/SimpleUpdateContext.java
+++ b/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/SimpleUpdateContext.java
@@ -17,7 +17,7 @@
 package com.blazebit.persistence.view.impl.update;
 
 import com.blazebit.persistence.view.impl.EntityViewManagerImpl;
-import com.blazebit.persistence.view.impl.tx.TransactionSynchronizationStrategy;
+import com.blazebit.persistence.view.spi.TransactionAccess;
 import com.blazebit.persistence.view.impl.update.flush.PostFlushDeleter;
 
 import javax.persistence.EntityManager;
@@ -73,7 +73,7 @@ public class SimpleUpdateContext implements UpdateContext {
         return false;
     }
 
-    public TransactionSynchronizationStrategy getSynchronizationStrategy() {
+    public TransactionAccess getTransactionAccess() {
         return null;
     }
 

--- a/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/UpdateContext.java
+++ b/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/UpdateContext.java
@@ -17,7 +17,7 @@
 package com.blazebit.persistence.view.impl.update;
 
 import com.blazebit.persistence.view.impl.EntityViewManagerImpl;
-import com.blazebit.persistence.view.impl.tx.TransactionSynchronizationStrategy;
+import com.blazebit.persistence.view.spi.TransactionAccess;
 import com.blazebit.persistence.view.impl.update.flush.PostFlushDeleter;
 
 import javax.persistence.EntityManager;
@@ -44,7 +44,7 @@ public interface UpdateContext {
 
     public boolean isRemovedObject(Object value);
 
-    public TransactionSynchronizationStrategy getSynchronizationStrategy();
+    public TransactionAccess getTransactionAccess();
 
     public InitialStateResetter getInitialStateResetter();
 

--- a/integration/entity-view-spring/pom.xml
+++ b/integration/entity-view-spring/pom.xml
@@ -52,18 +52,24 @@
             <artifactId>spring-context</artifactId>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>org.hibernate.javax.persistence</groupId>
+            <artifactId>hibernate-jpa-2.1-api</artifactId>
+            <version>1.0.0.Final</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>javax.transaction</groupId>
+            <artifactId>javax.transaction-api</artifactId>
+            <version>${version.jta}</version>
+            <scope>provided</scope>
+        </dependency>
 
         <!-- Test Dependencies -->
 
         <dependency>
             <groupId>com.blazebit</groupId>
             <artifactId>blaze-persistence-entity-view-impl</artifactId>
-            <scope>test</scope>
-        </dependency>
-        <dependency>
-            <groupId>org.hibernate.javax.persistence</groupId>
-            <artifactId>hibernate-jpa-2.1-api</artifactId>
-            <version>1.0.0.Final</version>
             <scope>test</scope>
         </dependency>
         <dependency>
@@ -109,6 +115,7 @@
                                         exports com.blazebit.persistence.integration.view.spring;
                                         exports com.blazebit.persistence.integration.view.spring.impl to spring.beans;
                                         opens com.blazebit.persistence.integration.view.spring.impl to spring.core, spring.context;
+                                        provides com.blazebit.persistence.view.spi.TransactionAccessFactory with com.blazebit.persistence.integration.view.spring.impl.SpringTransactionAccessFactory;
                                     }
                                 </moduleInfoSource>
                             </module>

--- a/integration/entity-view-spring/src/main/java/com/blazebit/persistence/integration/view/spring/impl/SpringTransactionAccessFactory.java
+++ b/integration/entity-view-spring/src/main/java/com/blazebit/persistence/integration/view/spring/impl/SpringTransactionAccessFactory.java
@@ -1,0 +1,37 @@
+/*
+ * Copyright 2014 - 2019 Blazebit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.blazebit.persistence.integration.view.spring.impl;
+
+import com.blazebit.persistence.view.spi.TransactionAccess;
+import com.blazebit.persistence.view.spi.TransactionAccessFactory;
+import org.springframework.transaction.support.TransactionSynchronizationManager;
+
+import javax.persistence.EntityManager;
+
+/**
+ * @author Christian Beikov
+ * @since 1.4.0
+ */
+public class SpringTransactionAccessFactory implements TransactionAccessFactory {
+
+    @Override
+    public TransactionAccess createTransactionAccess(EntityManager entityManager) {
+        if (TransactionSynchronizationManager.isActualTransactionActive()) {
+            return SpringTransactionSynchronizationStrategy.INSTANCE;
+        }
+        return null;
+    }
+}

--- a/integration/entity-view-spring/src/main/java/com/blazebit/persistence/integration/view/spring/impl/SpringTransactionSynchronizationStrategy.java
+++ b/integration/entity-view-spring/src/main/java/com/blazebit/persistence/integration/view/spring/impl/SpringTransactionSynchronizationStrategy.java
@@ -1,0 +1,102 @@
+/*
+ * Copyright 2014 - 2019 Blazebit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.blazebit.persistence.integration.view.spring.impl;
+
+import com.blazebit.persistence.view.spi.TransactionAccess;
+import org.springframework.transaction.interceptor.TransactionInterceptor;
+import org.springframework.transaction.support.TransactionSynchronization;
+import org.springframework.transaction.support.TransactionSynchronizationManager;
+
+import javax.transaction.Synchronization;
+
+/**
+ *
+ * @author Christian Beikov
+ * @since 1.4.0
+ */
+public class SpringTransactionSynchronizationStrategy implements TransactionAccess {
+
+    public static final SpringTransactionSynchronizationStrategy INSTANCE = new SpringTransactionSynchronizationStrategy();
+
+    private SpringTransactionSynchronizationStrategy() {
+    }
+
+    @Override
+    public boolean isActive() {
+        return TransactionSynchronizationManager.isActualTransactionActive();
+    }
+
+    @Override
+    public void markRollbackOnly() {
+        TransactionInterceptor.currentTransactionStatus().setRollbackOnly();
+    }
+
+    @Override
+    public void registerSynchronization(Synchronization synchronization) {
+        TransactionSynchronizationManager.registerSynchronization(new TransactionSynchronizationWrapper(synchronization));
+    }
+
+    /**
+     *
+     * @author Christian Beikov
+     * @since 1.4.0
+     */
+    private static class TransactionSynchronizationWrapper implements TransactionSynchronization {
+
+        private final Synchronization synchronization;
+
+        public TransactionSynchronizationWrapper(Synchronization synchronization) {
+            this.synchronization = synchronization;
+        }
+
+        @Override
+        public void suspend() {
+            // No-op
+        }
+
+        @Override
+        public void resume() {
+            // No-op
+        }
+
+        @Override
+        public void flush() {
+            // No-op
+        }
+
+        @Override
+        public void beforeCommit(boolean readOnly) {
+            // No-op
+        }
+
+        @Override
+        public void beforeCompletion() {
+            synchronization.beforeCompletion();
+        }
+
+        @Override
+        public void afterCommit() {
+            // No-op
+        }
+
+        @Override
+        public void afterCompletion(int status) {
+            synchronization.afterCompletion(status);
+        }
+    }
+
+}

--- a/integration/entity-view-spring/src/main/resources/META-INF/services/com.blazebit.persistence.view.spi.TransactionAccessFactory
+++ b/integration/entity-view-spring/src/main/resources/META-INF/services/com.blazebit.persistence.view.spi.TransactionAccessFactory
@@ -1,0 +1,1 @@
+com.blazebit.persistence.integration.view.spring.impl.SpringTransactionAccessFactory

